research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+    params: hyperparams of the model.
+  """
+  # Forge an initial checkpoint with the name that subsequent Estimator will
+  # expect to find.
+  estimator_initial_checkpoint_name = 'model.ckpt-1'
+  save_file = os.path.join(working_dir,
+                           estimator_initial_checkpoint_name)
+  sess = tf.Session(graph=tf.Graph())
+  with sess.graph.as_default():
+    input_features, labels = get_inference_input()
+    dualnet_model.model_fn(
+        input_features, labels, tf.estimator.ModeKeys.PREDICT, params)
+    sess.run(tf.global_variables_initializer())
+    tf.train.Saver().save(sess, save_file)
+
+
+def export_model(working_dir, model_path):
+  """Take the latest checkpoint and export it to model_path for selfplay.
+
+  Assumes that all relevant model files are prefixed by the same name.
+  (For example, foo.index, foo.meta and foo.data-00000-of-00001).
+
+  Args:
+    working_dir: The directory where tf.estimator keeps its checkpoints.
+    model_path: Either a local path or a gs:// path to export model to.
+  """
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params='ignored')
+  latest_checkpoint = estimator.latest_checkpoint()
+  all_checkpoint_files = tf.gfile.Glob(latest_checkpoint + '*')
+  for filename in all_checkpoint_files:
+    suffix = filename.partition(latest_checkpoint)[2]
+    destination_path = model_path + suffix
+    tf.gfile.Copy(filename, destination_path)
+
+
+def train(working_dir, tf_records, generation_num, params):
+  """Train the model for a specific generation.
+
+  Args:
+    working_dir: The model working directory to save model parameters,
+      drop logs, checkpoints, and so on.
+    tf_records: A list of tf_record filenames for training input.
+    generation_num: The generation to be trained.
+    params: hyperparams of the model.
+  """
+  assert generation_num > 0, 'Model 0 is random weights'
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params=params)
+  max_steps = (generation_num * params.examples_per_generation
+               // params.batch_size)
+  update_ratio_hook = model_utils.UpdateRatioSessionHook(working_dir)
+
+  def input_fn():
+    return preprocessing.get_input_tensors(params.batch_size, tf_records)
+  estimator.train(
+      input_fn, hooks=[update_ratio_hook, tf.train.ProfilerHook(save_secs=600)],
+      max_steps=max_steps)
+
+
+def validate(working_dir, tf_records, checkpoint_name, params):
+  """Perform model validation on the hold out data.
+
+  Args:
+    working_dir: The model working directory.
+    tf_records: A list of tf_records filenames for holdout data.
+    checkpoint_name: The checkpoint used to for evaluation.
+    params: hyperparams of the model.
+  """
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params=params)
+  if checkpoint_name is None:
+    checkpoint_name = estimator.latest_checkpoint()
```

Comment:
This doesn't seem to be used? (Or else I'm seeing things after 2+ hours of code review :) )

Reply:
You are right, it always evaluate the latest checkpoint in current implementation. The original code can provide validation to any checkpoint, here I just simplify it to the latest one. Will remove it. Thanks!